### DISCUSSION TITLE
Next build programmatically

### DISCUSSION THREAD

**Discussion by leimonio** [vercel/next.js#51515] [Help] [answered]
### Summary

I'm building a tool on top of NextJS and I'd like to be able to trigger the build programmatically, in a similar way to the local development server, using next as a module. Is this supported?

### Example

I had in mind something like

```
const nextApp = next({...})
await nextApp.prepare();
await nextApp.build();
```

**Comment by icyJoseph** [comment, marked as answer]
Mmm something like this?

- https://github.com/vercel/next.js/blob/canary/packages/next/src/cli/next-build.ts
- https://github.com/vercel/next.js/blob/canary/packages/next/src/build/index.ts

The CLI `next-build`, implements, or rather, calls, the `build` function from the second link.

**Reply by leimonio** [reply]
Yes I think I tried that, ideally I'd like to pass next configuration as well, to define output directory etc., similarly to 'next' server. 

The 'build' function only receives the directory of the NextJS project root.

**Reply by icyJoseph** [reply]
But the build loads the config relative to the dir path... as it, it looks for `next.config.js`, so you kinda have to work it there instead. Note sure what you are up to, but it looks like the build function is a bit more convoluted.

Otherwise, you are basically looking into re-implementing the build function so that it becomes pure. Most notably the async work done here:

- https://github.com/vercel/next.js/blob/canary/packages/next/src/build/index.ts#L231

Which is nearly 3000 lines of code 🤔 but you wanna to target this https://github.com/vercel/next.js/blob/canary/packages/next/src/build/index.ts#L238, not sure if the NextBuildContext, can help, it looks like it is a read-layer more than anything else.

What's needed though, is to move the env file and config file loading, out of the build function.

**Reply by leimonio** [reply]
I'm getting this error after defining a `next.config.js` file, both when I start the server or build.

```
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
Error occurred while building the Next.js application: TypeError: Invalid host defined options
    at loadConfig (/Users/leimonio/dev/bit.dev/bit-next-app/node_modules/.pnpm/next@13.3.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/config.js:528:78)
    at async Span.traceAsyncFn (/Users/leimonio/dev/bit.dev/bit-next-app/node_modules/.pnpm/next@13.3.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/trace/trace.js:103:20)
    at async /Users/leimonio/dev/bit.dev/bit-next-app/node_modules/.pnpm/next@13.3.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/build/index.js:148:28
    at async Span.traceAsyncFn (/Users/leimonio/dev/bit.dev/bit-next-app/node_modules/.pnpm/next@13.3.4_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/trace/trace.js:103:20)
```

`next.config.js`
```
module.exports = {
  reactStrictMode: true,
  output: 'export',
  distDir: 'next-build',
  images: {
    unoptimized: true
  }
}
```

**Reply by icyJoseph** [reply]
right, and `next.config.js` is where? relative to the script runner?

**Reply by leimonio** [reply]
No the script runner is in a different path not related to the next app. Using next server to serve the development bundle is possible. I can read the `next.config.js` file and pass it to `next({ ..., conf })` but for `build` I cannot do that. For build I need to define the the output method and directory.

**Reply by leimonio** [reply]
Changing this `import()` to  `resolve()` solves the issue and is able to require the `next.config.js` as expected.
https://github.com/vercel/next.js/blob/canary/packages/next/src/server/config.ts#L734
